dask/array/enhanced_numpy.py

```diff
@@ -0,0 +1,79 @@
+from __future__ import absolute_import, division, print_function
+
+from functools import wraps
+from inspect import getargspec
+
+import numpy as np
+
+from ..utils import ignoring
+
+
+
+# A wrapper for functions that don't provide keepdims to ensure that they do.
+def keepdims_wrapper(a_callable):
+    if "keepdims" in getargspec(a_callable).args:
+        return a_callable
+
+    @wraps(a_callable)
+    def keepdims_wrapped_callable(x, axis=None, keepdims=None, *args, **kwargs):
+        r = a_callable(x, axis=axis, *args, **kwargs)
+
+        if not keepdims:
+            return r
+
+        axes = axis
+
+        if axes is None:
+            axes = range(x.ndim)
+
+        try:
+            len(axes)
```

Comment:
That sounds reasonable, but I try to avoid checks of a particular type unless absolutely necessary. I have instead tried to use a series of abstract base classes, which ensure membership tests are possible as that is the only feature used here. As a result, we don't rule out other types, which have the right behavior simply because they aren't a specific type.